---
date: 2024-09-02
tags:
  - Процедурне_програмування
  - Основи
---
У попередньому розділі ми визначили функцію як набір операторів, які виконуються послідовно. Хоча це, безумовно, вірно, таке визначення не дає розуміння того, чому функції є корисними. Давайте оновимо наше визначення: **Функція** - це багаторазова послідовність операторів, призначена для виконання певної роботи.

Ви вже знаєте, що кожна виконувана програма повинна мати функцію з іменем main (саме з неї програма починає виконання під час запуску). Однак, оскільки програми стають дедалі довшими, розміщувати весь код у функції main стає дедалі важче. Функції дають нам можливість розбити наші програми на невеликі модульні частини, які легше організувати, тестувати та використовувати. Більшість програм використовують багато функцій. Стандартна бібліотека C++ постачається з великою кількістю вже написаних функцій, які ви можете використовувати, але не менш поширеною практикою є написання власних функцій. Функції, які ви пишете самостійно, називаються **функціями користувача**.

Розглянемо випадок, який може трапитися у реальному житті: ви читаєте книгу, коли згадуєте, що вам потрібно зателефонувати. Ви робите закладку у книзі, телефонуєте, а коли закінчуєте розмову, повертаєтеся до місця закладки і продовжуєте читати книгу саме з того місця, на якому зупинилися.

Програми на C++ можуть працювати так само. Програма буде послідовно виконувати оператори всередині однієї функції, коли зустріне виклик функції. **Виклик функції** - це вираз, який вказує центральному процесору перервати поточну функцію і виконати іншу функцію. Процесор "ставить закладку" в поточній точці виконання, а потім викликає (виконує) функцію, названу у виклику функції. Коли викликана функція завершується, процесор повертається до точки, на якій була зроблена закладка, і продовжує виконання.

Функція, що ініціює виклик функції, називається **викликаючою**, а функція, що викликається, - **викликаною** або **викликаною функцією.**
## Приклад користувацької функції
Почнемо з найпростішого синтаксису для визначення користувацької функції. Протягом наступних кількох уроків усі користувацькі функції матимуть такий вигляд:

```cpp
returnType functionName() // This is the function header (tells the compiler about the existence of the function)
{
    // This is the function body (tells the compiler what the function does)
}
```

Перший рядок неформально називається заголовком функції, і він повідомляє компілятору про існування функції, її ім'я та деяку іншу інформацію.

Фігурні дужки та оператори між ними називаються **тілом функції**. Це місце, де містяться оператори, які визначають, що робить ваша функція.

Ось приклад програми, яка показує, як визначається і викликається нова функція:

```cpp
#include <iostream> // for std::cout

// Definition of user-defined function doPrint()
void doPrint() // doPrint() is the called function in this example
{
    std::cout << "In doPrint()\n";
}

// Definition of function main()
int main()
{
    std::cout << "Starting main()\n";
    doPrint(); // Interrupt main() by making a function call to doPrint().  main() is the caller.
    std::cout << "Ending main()\n"; // this statement is executed after doPrint() ends

    return 0;
}
```

Ця програма виводить наступний результат:

```cpp
Starting main()
In doPrint()
Ending main()
```

Ця програма починає виконання з вершини функції main, і перший рядок, який виконується, виводить `Starting main()`.

Другий рядок в main - це виклик функції doPrint. Ми викликаємо функцію doPrint, додаючи пару круглих дужок до імені функції, наприклад: `doPrint()`. Зверніть увагу, що якщо ви забудете дужки, ваша програма може не скомпілюватися (а якщо скомпілюється, то функція не буде викликана).

> [!caution] 
> Не забудьте включити круглі дужки () після імені функції при її виклику.

Оскільки було здійснено виклик функції, виконання операторів у main призупиняється, і виконання переходить до вершини викликаної функції doPrint. Перший (і єдиний) рядок у doPrint друкується `In doPrint()`. Коли doPrint завершується, виконання повертається назад до того, хто його викликав (тут: до функції main), і відновлюється з того місця, на якому воно зупинилося. Отже, наступний оператор, що виконується в main, виводить `Ending main()`.
## Виклик функцій більше одного разу
Однією з корисних речей у функціях є те, що їх можна викликати більше одного разу. Ось програма, яка це демонструє:

```cpp
#include <iostream> // for std::cout

void doPrint()
{
    std::cout << "In doPrint()\n";
}

// Definition of function main()
int main()
{
    std::cout << "Starting main()\n";
    doPrint(); // doPrint() called for the first time
    doPrint(); // doPrint() called for the second time
    std::cout << "Ending main()\n";

    return 0;
}
```

Ця програма виводить наступний результат:

```cpp
Starting main()
In doPrint()
In doPrint()
Ending main()
```

Оскільки doPrint викликається двічі з main, doPrint виконується двічі, і In `doPrint()` друкується двічі (по одному разу для кожного виклику).
## Функції можуть викликати функції, які викликають інші функції
Ви вже бачили, що функція main може викликати інші функції (наприклад, функцію doPrint у наведеному вище прикладі). Функції, які викликаються за допомогою main, також можуть викликати інші функції (а ті можуть викликати інші функції і т.д.). У наступній програмі функція main викликає функцію doA, яка викликає функцію doB:

```cpp
#include <iostream> // for std::cout

void doB()
{
    std::cout << "In doB()\n";
}

void doA()
{
    std::cout << "Starting doA()\n";

    doB();

    std::cout << "Ending doA()\n";
}

// Definition of function main()
int main()
{
    std::cout << "Starting main()\n";

    doA();

    std::cout << "Ending main()\n";

    return 0;
}
```

Ця програма виводить наступний результат:

```cpp
Starting main()
Starting doA()
In doB()
Ending doA()
Ending main()
```
## Вкладені функції не підтримуються
На відміну від деяких інших мов програмування, у C++ функції не можуть бути визначені всередині інших функцій. Наступна програма не є легальною:

```cpp
#include <iostream>

int main()
{
    void foo() // Illegal: this function is nested inside function main()
    {
        std::cout << "foo!\n";
    }

    foo(); // function call to foo()
    return 0;
}
```

Правильний спосіб написання вищенаведеної програми наступний:

```cpp
#include <iostream>

void foo() // no longer inside of main()
{
    std::cout << "foo!\n";
}

int main()
{
    foo();
    return 0;
}
```